### PR TITLE
[OP#48106] Add a reference provider and a reference widget for work packages

### PR DESCRIPTION
**Pull request by julien-nc**
This is just the structure, the provider and the widget do nothing at the moment.
This requires https://github.com/nextcloud/integration_openproject/pull/380 as the widget registration function appeared recently in `@nextcloud/vue` and the latest `@nextcloud/webpack-vue-config` is required.

The registration of the reference provider and of the widget only happen on NC >= 26.

The reference system is documented there: https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/reference.html

The reference provider has 2 goals:
* Add a smart picker provider so users can get work package links
* Resolve and render work package link previews

### The smart picker

There are 2 ways to implement a smart picker provider. This PR implements the first one.

1. Just use a unified search provider and let the picker show a generic rendering for the provider (like for other apps like integration_github, integration_gitlab, collectives, deck, tables...). In this case there is nothing to implement on the frontend side. Users can perform a basic search and the picker result is the "resourceUrl" of the selected search result entry.
2. It is possible to register a custom picker component to have a custom interface. For example, this is done in integration_openstreetmap (it shows a map to get a location link), in integration_giphy (it shows a GIF picker with a grid view and infinite scrolling) etc...

We can start by implementing the first one and see if it makes sense to have a custom picker later. It would allow us to add search options for example.

### The link previews (reference widgets)

If a reference provider recognizes a link that it can handle, it resolves it and provides information for the frontend.
If only the generic title, description and image are set, the default rendering will be used.
The rendering can be customized if a rich object is set and a widget component is registered for this rich object type.
This is done in this PR.

Related workpackage: https://community.openproject.org/projects/nextcloud-integration/work_packages/48106/activity

### CODE REVIEW

**Comment by SwikritiT** [comment]
https://github.com/nextcloud/integration_openproject/pull/380 is merged now, I'll rebase and look around.

File: lib/Reference/WorkPackageReferenceProvider.php [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+		return null;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function matchReference(string $referenceText): bool {
+		if ($this->userId !== null) {
+			$linkPreviewEnabled = $this->config->getUserValue($this->userId, Application::APP_ID, 'default_enable_unified_search', '0') === '1';
```
Comment (SwikritiT) [inline lib/Reference/WorkPackageReferenceProvider.php:123]: reminder to self to change these and  add a migration step to add this key
```suggestion
			$linkPreviewEnabled = $this->config->getUserValue($this->userId, Application::APP_ID, ''link_preview_enabled'', '0') === '1';
```

File: lib/Reference/WorkPackageReferenceProvider.php [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+			if (!$linkPreviewEnabled) {
+				return false;
+			}
+		}
+		$adminLinkPreviewEnabled = $this->config->getAppValue(Application::APP_ID, 'default_enable_unified_search', '0') === '1';
```
Comment (SwikritiT) [inline lib/Reference/WorkPackageReferenceProvider.php:128]:
```suggestion
		$adminLinkPreviewEnabled = $this->config->getAppValue(Application::APP_ID, ''link_preview_enabled'', '0') === '1';
```

File: phpstan.neon [hunk 3]
```diff
@@ -7,3 +7,5 @@ parameters:
         - bootstrap.php
     checkGenericClassInNonGenericObjectType: false
     reportUnmatchedIgnoredErrors: false
+    excludePaths:
+    	-  './lib/Reference/WorkPackageReferenceProvider.php'
```
Comment (SwikritiT) [inline phpstan.neon:11]: Phpstan will throw `Reflection error: OCP\Collaboration\Reference\ISearchableReferenceProvider not found.` in nextcloud server version lower that 25, as this is not available in lower versions, so exclude the path

**Review by individual-it — APPROVED** [review]
Looks good to me, let's merge it when CI passes

File: .php-cs-fixer.dist.php [hunk 4]
```diff
@@ -15,5 +15,6 @@
 	->notPath('src')
 	->notPath('vendor')
 	->notPath('server')
+	->notPath('lib/Reference/WorkPackageReferenceProvider.php')
```
Comment (SwikritiT) [inline .php-cs-fixer.dist.php:18]: similar to this https://github.com/nextcloud/integration_openproject/pull/399/files#r1230459632 but with `cs:fixer`

**Pull request merged**